components/brave_wallet/browser/keyring_service.cc

```diff
@@ -1726,6 +1687,84 @@ void KeyringService::Reset(bool notify_observer) {
   }
 }
 
+void KeyringService::MaybeMigratePBKDF2Iterations(const std::string& password) {
+  if (!prefs_->GetBoolean(kBraveWalletKeyringsNeedPasswordsMigration)) {
+    return;
+  }
+
+  // Pref is supposed to be set only as false.
+  DCHECK(!prefs_->HasPrefPath(kBraveWalletKeyringsNeedPasswordsMigration));
+
+  for (auto* keyring_id :
+       {mojom::kDefaultKeyringId, mojom::kFilecoinKeyringId,
+        mojom::kFilecoinTestnetKeyringId, mojom::kSolanaKeyringId}) {
+    auto legacy_encrypted_mnemonic =
+        GetPrefInBytesForKeyring(prefs_, kEncryptedMnemonic, keyring_id);
+    auto nonce =
+        GetPrefInBytesForKeyring(prefs_, kPasswordEncryptorNonce, keyring_id);
+    auto salt =
+        GetPrefInBytesForKeyring(prefs_, kPasswordEncryptorSalt, keyring_id);
+
+    if (!legacy_encrypted_mnemonic || !nonce || !salt) {
+      continue;
+    }
+
+    auto legacy_encryptor = PasswordEncryptor::DeriveKeyFromPasswordUsingPbkdf2(
+        password, *salt, kPbkdf2IterationsLegacy, kPbkdf2KeySize);
+    if (!legacy_encryptor)
+      continue;
+
+    auto mnemonic =
+        legacy_encryptor->Decrypt(*legacy_encrypted_mnemonic, *nonce);
+    if (!mnemonic)
+      continue;
+
+    auto encryptor = PasswordEncryptor::DeriveKeyFromPasswordUsingPbkdf2(
+        password, *salt, GetkPbkdf2Iterations(), kPbkdf2KeySize);
+    if (!encryptor)
+      continue;
+
+    SetPrefInBytesForKeyring(
+        prefs_, kEncryptedMnemonic,
+        encryptor->Encrypt(base::make_span(*mnemonic), *nonce), keyring_id);
```

Comment:
We need to be generating a new nonce used to encrypt here rather than reusing the one that was used previously to encrypt the keyring. Otherwise there's a crypto bug in this that would allow an attacker to combine the old ciphertext with the new ciphertext in a way that could reveal the XOR'd plaintext of the two messages.

Reply:
++ this is the only place in this PR we need to use a new nonce for. See this thread for details: https://github.com/brave/brave-core/pull/14523/files#r941156147

Reply:
generating new nonce here
https://github.com/brave/brave-core/pull/14523/commits/bd0e5d57772786a7be7a3b7e51d08caf9c16ea23